scio-parquet/src/main/scala/com/spotify/scio/parquet/read/ParquetRead.scala

```diff
@@ -43,8 +57,163 @@ object ParquetRead {
-    ParDo.of(sdf)
+    val tfx: PTransform[PCollection[_ <: ReadableFile], PCollection[R]] = ParDo.of(sdf)
+
+    tfx.asInstanceOf[PTransform[PCollection[ReadableFile], PCollection[R]]]
+  }
+
+  /**
+   * A ReadFiles implementation that reads Parquet file(s) into Scala case classes of type T
+   *
+   * @param predicate
+   *   a Parquet [[FilterPredicate]] predicate, if desired
+   * @param conf
+   *   a Parquet [[Configuration]], if desired
+   */
+  def readTyped[T: ClassTag: ParquetType](
+    predicate: FilterPredicate = null,
+    conf: Configuration = null
+  ): PTransform[PCollection[ReadableFile], PCollection[T]] = readTyped[T, T](
+    identity,
+    predicate,
+    conf
+  )
+
+  /**
+   * A ReadFiles implementation that reads Parquet file(s) into Scala case classes of type R, where
+   * R is mapped from type T
+   *
+   * @param projectionFn
+   *   a function mapping T => R
+   * @param predicate
+   *   a Parquet [[FilterApi]] predicate
+   * @param conf
+   *   a Parquet [[Configuration]]
+   */
+  def readTyped[T: ClassTag: ParquetType, R](
+    projectionFn: T => R,
+    predicate: FilterPredicate,
+    conf: Configuration
+  ): PTransform[PCollection[ReadableFile], PCollection[R]] = {
+    val configuration = ParquetConfiguration.ofNullable(conf)
+    Option(predicate).foreach(p => ParquetInputFormat.setFilterPredicate(configuration, p))
+
+    val cleanedFn = Functions.serializableFn(ClosureCleaner.clean(projectionFn))
+    readFiles(ReadSupportFactory.typed[T], new SerializableConfiguration(configuration), cleanedFn)
+  }
+
+  /**
+   * A ReadFiles implementation that reads Parquet file(s) into Avro [[GenericRecord]]s using the
+   * supplied schema
+   *
+   * @param schema
+   *   The Avro [[Schema]] to use for Parquet reads; can be a projection of the full file schema
+   * @param predicate
+   *   a Parquet [[FilterPredicate]] predicate, if desired
+   * @param conf
+   *   a Parquet [[Configuration]], if desired
+   */
+  def readAvroGenericRecordFiles(
+    schema: Schema,
+    predicate: FilterPredicate = null,
+    conf: Configuration = null
+  ): PTransform[PCollection[ReadableFile], PCollection[GenericRecord]] =
+    readAvroGenericRecordFiles[GenericRecord](schema, identity, predicate, conf)
+
+  /**
+   * A ReadFiles implementation that reads Parquet file(s) into Avro [[GenericRecord]]s using the
+   * supplied schema, then applies a mapping function to convert the Avro records into type T
+   *
+   * @param schema
+   *   The Avro [[Schema]] to use for Parquet reads; can be a projection of the full file schema
+   * @param projectionFn
+   *   a function mapping [[GenericRecord]] => T
+   * @param predicate
+   *   a Parquet [[FilterPredicate]] predicate
+   * @param conf
+   *   a Parquet [[Configuration]]
+   */
+  def readAvroGenericRecordFiles[T](
+    schema: Schema,
+    projectionFn: GenericRecord => T,
+    predicate: FilterPredicate,
+    conf: Configuration
```

Comment:
Should we create method overload to avoid setting those at null ?

Reply:
I've been thinking about that -- my concern is that overloading might make things more confusing from user POV. Say  we have:

````scala
def readAvro[T <: SpecificRecord: ClassTag](): PTransform[PCollection[ReadableFile], PCollection[T]] =
    readAvro[T](null, null, null)

def readAvro[T <: SpecificRecord: ClassTag](
    projection: Schema,
    predicate: FilterPredicate,
    conf: Configuration
  ): PTransform[PCollection[ReadableFile], PCollection[T]] = ...
```

In this case, users can either pass 0 arguments, or all 3. But I think it's somewhat common use case to pass in only a projection, or only a predicate -- and that would be more difficult for the user if we went with the overloaded API.

Reply:
I meant we can create overloads for all possibilities